### PR TITLE
Feature/RX-1704

### PR DESCRIPTION
**Pull request by luig0**
## Summary

Add WordCount component + testing

## What changed in the code?

Made modifications to App.js and index.js to pass down sample snippets for testing (in the absence of passed in snippets). Added WordCount component to Editor component.
Added files to src/Ckeditor/: -WordCount.js
* WordCount.test.js
* wordcount.css

## Test coverage

Added unit testing for counts.

#### Gaps in Coverage (i.e. missing tests)

No integration testing yet.
I copy/pasted a function from the CKE word-count plugin which converts CKE elements into plain text.
* I did not write tests for this function because I was not able to obtain a reliable string of the output of editor.model.document.getRoot()
* I was not able to immediately find any existing tests in CKE code for testing this functionality

#### Additional Testing Notes

I wrote Jest tests for various scenarios of word/char count calculation. Included tests for snippet calc and punctuation handling.

## Anything else before merging (i.e. why is this on hold)?

- [x] Update for plugins file structure
- [x] Update word count filenames to snake case
- [x] Change sample snippet values so that they are a different number of words than the associated key
- [x] Add test case for character count limit
- [x] use `env.features.isRegExpUnicodePropertySupported`
- [x] debounce keystrokes before word count update
- [ ] Verify functionality with story/RX-1677

### CODE REVIEW

**Comment by giarcjc** [comment]
I think it would be good to move `src/Ckeditor/WordCount.js`, `src/Ckeditor/WordCount.test.js` and `src/Ckeditor/wordcount.css` to be in a `src/wordcount` dir.  Also, we haven't really discussed naming conventions for this lib, but it looks like the convention that the CKeditor team follows is all lowercase, no spaces or underscores (e.g. `node_modules/@ckeditor/ckeditor5-utils/src/areconnectedthroughproperties.js`).  I think we should probably try to be consistent with that although certainly worth a discussion.

File: src/Ckeditor/WordCount.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+ * - Counts are adjusted for snippets
+ * - Snippets replacement algo is a slimmed down version of that which is used
+ *   in Ombud Rx with Slate
+ * - Counting algo is a slimmed down version of that which is provided by the
+ *   CKEditor word count plugin. The CKE plugin is not used directly because we
+ *   needed to also count the words and characters from snippets prior to
+ *   rendering the final count.
+ * - Reference: https://www.npmjs.com/package/@ckeditor/ckeditor5-word-count
+ * - Reference: https://ckeditor.com/docs/ckeditor5/latest/api/module_word-count_wordcount-WordCount.html
+ */
```
Comment (giarcjc) [inline src/Ckeditor/WordCount.js:14]: +1 for helpful usage comments

File: src/Ckeditor/WordCount.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    // Firefox throws a SyntaxError when the group is unsupported.
+  }
+  return isSupported;
+}
+
+/**
+ * Convert editor elements into plain text
+ * @param   {RootElement} element The root element to be parsed into plain text
+ * @return  {String}              The element tree converted to plain text
+ */
```
Comment (giarcjc) [inline src/Ckeditor/WordCount.js:64]: Another discussion item for us to coalesce on as a team.  Do we want to standardize on using JSDoc in the project?

Comment (luig0) [reply src/Ckeditor/WordCount.js:64]: Hey yes let's please discuss this at today's call. I followed a page I found online, but definitely want to conform if there has been any discussion here.

**Comment by luig0** [comment]
> I think it would be good to move `src/Ckeditor/WordCount.js`, `src/Ckeditor/WordCount.test.js` and `src/Ckeditor/wordcount.css` to be in a `src/wordcount` dir. Also, we haven't really discussed naming conventions for this lib, but it looks like the convention that the CKeditor team follows is all lowercase, no spaces or underscores (e.g. `node_modules/@ckeditor/ckeditor5-utils/src/areconnectedthroughproperties.js`). I think we should probably try to be consistent with that although certainly worth a discussion.

- Agree on putting word count files into a separate directory. Original discussion was to put it here, but I think we hadn't considered at the time yet that there might be a `.css` and `.test.js`
- Also agree on establishing naming convention and being consistent! Will bring this up at today's call as well

**Comment by ejwill04** [comment]
> > I think it would be good to move `src/Ckeditor/WordCount.js`, `src/Ckeditor/WordCount.test.js` and `src/Ckeditor/wordcount.css` to be in a `src/wordcount` dir. Also, we haven't really discussed naming conventions for this lib, but it looks like the convention that the CKeditor team follows is all lowercase, no spaces or underscores (e.g. `node_modules/@ckeditor/ckeditor5-utils/src/areconnectedthroughproperties.js`). I think we should probably try to be consistent with that although certainly worth a discussion.
> 
> * Agree on putting word count files into a separate directory. Original discussion was to put it here, but I think we hadn't considered at the time yet that there might be a `.css` and `.test.js`
> * Also agree on establishing naming convention and being consistent! Will bring this up at today's call as well

I agree with both of you about adding a folder structure for plugins.  Let's just make the main output file for each plugin an `index.js` file so we can just reference the folder name. 

Regarding naming, I'd be in favor of either snake_case or TitleCase for the file names.

**Comment by ejwill04** [comment]
@luig0 - Do you mind updating this branch for the new file structure we discussed?

**Comment by ejwill04** [comment]
Putting the hold tag on this ticket until we get RX-1677 merged in.

File: src/Ckeditor/WordCount.test.js [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+  it('should render the correct count of words and characters when no snippets are present', () => {
+    editorText = 'The quick brown fox jumps over the lazy dog.';
+    act(() => {
+      render(<WordCount editor={mockEditor} snippets={snippets} />, container);
+    });
+    expect(container.textContent).toContain('9 words');
+    expect(container.textContent).toContain('44 characters');
+  });
+
+  it('should render the correct count of words and characters after snippet replacement', () => {
```
Comment (ejwill04) [inline src/Ckeditor/WordCount.test.js:73]: I love these tests!  We'll need to update the logic for snippets to register the word count length in the value when the snippet is present.  

snippets: {
  document_name: "This super cool rad name"
}

For instance, `{{document_name}}` is five words here.  
However, if the snippet is not found (`{{fake_one}}`) is two words

**Comment by luig0** [comment]
> @luig0 - Do you mind updating this branch for the new file structure we discussed?

Added task list to initial comment:
- Update for plugins file structure
- Update for snake case

File: src/Ckeditor/WordCount.test.js (hunk 3 above)
Comment (luig0) [reply src/Ckeditor/WordCount.test.js:73]: Think this is how it is currently done. Will double check my tests.

Comment (ejwill04) [reply src/Ckeditor/WordCount.test.js:73]: The test is correct.  No action is needed.

**Comment by luig0** [comment]
Closing PR as this code is being moved to Rx.

**Pull request closed without merging**